katdal/caldata.py

```diff
@@ -0,0 +1,400 @@
+                           for prod in katdal_obj.corr_products]
+        self._cp_lookup = np.asarray(self._cp_lookup, dtype=int)
+        # interpolation function over available solutions
+        for key in self._cal_solns.keys():
+            try:
+                self._cal_solns[key]['interp'] = _get_cal_product('cal_product_' + key,
+                                                                  katdal_obj,
+                                                                  kind=self._cal_solns[key]['kind'],
+                                                                  )
+            except:  # no delay cal solution from telstate
+                # should raise a warning
+                raise
+        self._cal_interp(katdal_obj.timestamps)
+
+    def _cal_interp(self, timestamps):
+        """Interpolate values between calculated timestamps"""
+        # Interpolate the calibration solutions for the selected range
+        for key in self._cal_solns.keys():
+            if self._cal_solns[key]['interp'] is not None:
+                # interpolate over values, or repeat single value
+                # expected size <ts><ch><pol><ant>
+                if hasattr(self._cal_solns[key]['interp'], '__call__'):
+                    solns = self._cal_solns[key]['interp'](timestamps)
+                else:
+                    solns = np.broadcast_to(self._cal_solns[key]['interp'], timestamps.shape)
+                if key == 'K':
+                    solns = np.exp(solns * self._delay_to_phase)
+                self._cal_solns[key]['solns'] = da.from_array(solns,
+                                                              chunks=(self.ts_chunk_size, self.ch_chunk_size, 1, 1),
+                                                              )
+
+
+def applycal(katdal_obj):
+    """
+    Apply the K, B, G solutions to visibilities at provided timestamps.
+    Optionally recompute the weights as well.
+
+    Returns
+    -------
+    katdal_obj: containing vis and weights (optional) with cal solns applied.
+    """
+
+    katdal_obj.cal_coeffs = []
+    katdal_obj.weight_coeffs = []
+
+    def _cal_calc(katdal_obj):
+        """Calculate calibration and weight coefficients"""
+
+        cal_obj = _cal_setup(katdal_obj)
+        _bls_len = katdal_obj._corrprod_keep.sum()
+        _chan_len = katdal_obj._freq_keep.sum()
+        _time_len = katdal_obj._time_keep.sum()
+
+        if _bls_len != len(cal_obj._cp_lookup):
+            raise CalibrationValueError('Shape mismatch between correlation products.')
+        if _chan_len != len(cal_obj._data_channel_freqs):
+            raise CalibrationValueError('Shape mismatch in frequency axis.')
+        if _time_len != len(katdal_obj.timestamps):
+            raise CalibrationValueError('Shape mismatch in timestamps.')
+
+        # calibrate visibilities
+        K = cal_obj._cal_solns['K']['solns']
+        B = cal_obj._cal_solns['B']['solns']
+        G = cal_obj._cal_solns['G']['solns']
```

Comment:
Rename the attributes you're accessing from outside to exclude the leading underscore.